### PR TITLE
Adding "TF Serving" Keras code example

### PR DESCRIPTION
**Pull request by dimitreOliveira**
Hi as discussed a while ago at [this thread](https://github.com/keras-team/keras/issues/17046), I am making this PR to add a Keras code example on how to serve models using TF Serving.

A few notes on this notebook:
- The goal here is to provide a simple example of teaching good practices and how to get started with TensorFlow/Keras and TFServing.
- I was not able to fully reproduce the TFServing behavior dues to the constraints of Colab and the scripts that generates the `tutobooks`.
  - Colab does not support Docker, so I could not use the recommended Docker setup for TFServing.
  - The scripts that generate the `tutobooks` has some issues with running Docker services in the background, I also was not able to run background processes for a similar reason.

Due to those issues running the TFServing service I created the notebook with all the Python code that I was able to run there, the rest of the code (TFServing-related) I had to write on Markdown text showing the code input and its expected outputs. I referenced a Colab notebook that I created and is able to run everything fully.

I hope that those issues are not a problem and we can add this example, it took some time to get everything together, let me know if you have any suggestions.

### CODE REVIEW

File: examples/keras_recipes/tf_serving.py [hunk 1]
```diff
@@ -0,0 +1,445 @@
+"""
+Title: Serving TensorFlow models with TFServing
+Author: [Dimitre Oliveira](https://www.linkedin.com/in/dimitre-oliveira-7a1a0113a/)
+Date created: 2022/12/27
+Last modified: 2022/12/27
+Description: How to serve TensorFlow models with TensorFlow Serving.
+Accelerator: GPU
```
Comment (fchollet) [inline examples/keras_recipes/tf_serving.py:7]: Since most of the code is not runnable, is this necessary?

File: examples/keras_recipes/tf_serving.py [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+"""
+
+"""
+# Introduction
```
Comment (fchollet) [inline examples/keras_recipes/tf_serving.py:11]: Please convert all section titles to `##` for proper rendering

File: examples/keras_recipes/tf_serving.py [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+
+Once you build your machine learning models, the next likely step is to serve them. You may want to do that by exposing your model as an endpoint service, there are many frameworks that you can use to do that, but the TensorFlow ecosystem has its own solution called [TensorFlow Serving](https://www.tensorflow.org/tfx/guide/serving).
```
Comment (fchollet) [inline examples/keras_recipes/tf_serving.py:13]: Please shorten text lines to within 80 chars (except links)

File: examples/keras_recipes/tf_serving.py [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+- Pixel values scaled to the `[-1, 1]` range
+- Images with the shape of `(224, 224, 3)` meaning `(height, width, channels)`
+
+We can do all of that with the following function:
+"""
+
+
+def preprocess(image, mean=0.5, std=0.5, shape=(224, 224)):
+    """
+    Scale, normalize and resizes images.
```
Comment (fchollet) [inline examples/keras_recipes/tf_serving.py:68]: The docstring description should be on the first line after `"""`

File: examples/keras_recipes/tf_serving.py [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+
+- Pixel values normalized to the `[0, 1]` range
+- Pixel values scaled to the `[-1, 1]` range
+- Images with the shape of `(224, 224, 3)` meaning `(height, width, channels)`
+
+We can do all of that with the following function:
+"""
+
+
+def preprocess(image, mean=0.5, std=0.5, shape=(224, 224)):
```
Comment (fchollet) [inline examples/keras_recipes/tf_serving.py:66]: For a Keras application, you should use `tf.keras.applications.mobile_net.preprocess_input`

File: examples/keras_recipes/tf_serving.py [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+imagenet_labels_url = (
+    "https://storage.googleapis.com/download.tensorflow.org/data/ImageNetLabels.txt"
+)
+response = requests.get(imagenet_labels_url)
+labels = [x for x in response.text.split("\n") if x != ""][1:]  # skiping backgroung
+# Convert the labels to the TensorFlow data format
+tf_labels = tf.constant(labels, dtype=tf.string)
+
+
+def postprocess(prediction, labels=tf_labels):
```
Comment (fchollet) [inline examples/keras_recipes/tf_serving.py:95]: You should use `tf.keras.applications.mobile_net.decode_predictions`

File: examples/keras_recipes/tf_serving.py [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+print(f"Model output shape: {model_outputs.shape}")
+print(f"Predicted class: {postprocess(model_outputs)}")
+
+"""
+# Save the model
+
+To load our trained model into TensorFlow Serving we first need to save it in [SavedModel](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/saved_model) format. This will create a protobuf file in a well-defined directory hierarchy, and will include a version number. [TensorFlow Serving](https://www.tensorflow.org/tfx/guide/serving) allows us to select which version of a model, or "servable" we want to use when we make inference requests. Each version will be exported to a different sub-directory under the given path.
+"""
+
+base_dir = tempfile.gettempdir()
```
Comment (fchollet) [inline examples/keras_recipes/tf_serving.py:139]: Why use a temp dir?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! This is a very useful tutorial.

**Comment by dimitreOliveira** [comment]
Hi there @fchollet I am glad that you liked this code, and thank you very much for the review and suggestions.

I have addressed all of the changes that you requested with exception of the `preprocess_input` and `decode_predictions` part, the reason is that I explicitly wanted to show the user what is usually required at those steps (preprocessing and postprocessing) so that they could understand one of the main advantages of the custom signatures, and embedding processing steps into the model graph in general.
Still, I added a section mentioning the usage of `preprocess_input` and `decode_predictions` so that the reader is aware of the best practices, let me know if this works for you, otherwise, I could just use the functions and use some text to describe what is going on under the hood.

**Review by fchollet — APPROVED** [review]
Thanks for the updates! Please add the generated files. Do note that I've pushed some minor copyedits, please pull them first.

**Comment by fchollet** [comment]
> BTW let me take this opportunity to wish you a happy new year!

Same to you!

**Comment by dimitreOliveira** [comment]
Thank you very much for the edits, I am very happy with the end results! I have just added the generated files.

Hope we can collaborate again in the near future.

On this same note, looking around I have not seen too many good practices from Keras/TF users regarding embedding preprocessing/postprocessing logic into the models, either using a custom serving signature or just concatenating those functions into the model (like we can do with the Keras layers).
Do you think that another Keras example on this topic could be useful? your code [Structured data classification from scratch](https://keras.io/examples/structured_data/structured_data_classification_from_scratch/) and the one that I am submitting already touches this topic. I don't want to be redundant, but I have a blog post written on the subject that discusses the trade-offs of the approach "[Shortening model deployment with TensorFlow
](https://towardsdatascience.com/shortening-model-deployment-with-tensorflow-d5a11044d0d)".

In summary, the idea would be to discuss the benefits of embedding preprocessing/postprocessing into the model and the code on how to do it with custom signatures or combining TensorFlow code with the models.

**Comment by fchollet** [comment]
> Do you think that another Keras example on this topic could be useful?

Definitely, that's a topic that comes up often and an explainer on this would be super helpful to many.

**Comment by dimitreOliveira** [comment]
Awesome, I will work on that topic in the future.

Regarding this PR, is there anything else needed before we merge it?

**Review by fchollet — APPROVED** [review]
Thanks for the great contribution! 👍
